mokapot/tabular_data.py

```diff
@@ -46,10 +50,121 @@ def get_score_column_type(suffix):
         return pa.float64()
     elif suffix in CSV_SUFFIXES:
         return "float"
+    elif suffix in PIN_SUFFIXES:
+        return "float"
     else:
         raise ValueError(f"Suffix '{suffix}' does not match expected formats")
 
 
+def is_traditional_pin(path: Path) -> bool:
+    """Check if the PIN file is a traditional PIN file.
+
+    The traditional PIN file uses tabs both as field delimiters
+    and as the separator for multiple proteins in the last column.
+
+    So it can be identified if:
+    1. The last column name is `protein(s)?`.
+    2. The header is tab delimited.
+    3. The rest of the file is tab delimited.
+    4. The number of delimiters in the other rows is >= number of columns.
+
+    Parameters
+    ----------
+    path : Path
+        The path to the PIN file.
+
+    Returns
+    -------
+    bool
+        True if the PIN file is a traditional PIN file, False otherwise.
+
+    Raises
+    ------
+    ValueError
+        If the PIN file is not a PIN file. (or is corrupted)
+    """
+    with open(path) as f:
+        header = f.readline().strip()
+        header = header.split("\t")
+        if len(header) == 1:
+            raise ValueError(f"File '{path}' file is not a PIN file.")
+
+        if not header[-1].lower().startswith("protein"):
+            raise ValueError(
+                f"File '{path}' is not a PIN file "
+                f"(last column '{header[-1]}' is not protein)."
+            )
+
+        num_fields = len(header)
+        for line in f:
+            line = line.strip()
+            if line.startswith("#") or line.startswith("DefaultDirection"):
+                continue
+
+            local_num_fields = len(line.split("\t"))
+            if local_num_fields < num_fields:
+                raise ValueError(
+                    f"File '{path}' is not a PIN file "
+                    "(number of fields is less than number of columns)."
+                )
+
+            if local_num_fields > num_fields:
+                return True
+
+        return False
+
+
+def read_traditional_pin(path) -> pd.DataFrame:
+    """Reads the file in memory and bundles the proteins.
+
+    The PIN file is assumed to be a traditional PIN file.
+    The PIN file is read in memory and the proteins are bundled into a single
+    column.
+
+    Parameters
+    ----------
+    path : Path
+        The path to the PIN file.
+
+    Returns
+    -------
+    pd.DataFrame
+        The PIN file as a pandas DataFrame.
+    """
+    header_names = None
+    num_cols = None
+
+    out_lines = []
```

Comment:
This accumulating in-memory breaks the assumption that we can stream independent of input file size.
For example, running Mokapot on a PIN file from 10x Astral files would consume a lot of memory.

Reply:
I agree, but my general thought was ... if you care about that you can reformat the file to any of the formats that support streaming (fix the separator, rename to tsv/csv in the simplest of cases). I could add a line where if more than 50k lines are parsed, it notifies the user that the file is large and a conversion might be a good idea (or to use comet with the separator parameter ... I cannot think of many other search engines right now that keep the protein separator as tabs).

The alternatives I thought were to 
1. have a temp file where the lines are dumped and then read with any of the streaming alternatives. But since this only really matters for large files, I didn't like the idea of duplicating a massive file that might or might not get cleaned correctly.
2. Have a non-temp file generated in the same dir location (It felt a bit intrusive to clutter the user's directory).
3. Error out and point the user to a sub-command that handles the conversion (very non-backwards compatible).

do you guy use the streaming features with plain text .pin files?

Reply:
We do not use PIN files as inputs internally anymore, so this should be fine. I was just concerned that the claim that Mokapot is all streamed could then not be made for traditional pins. My pin_to_tsv conversion is streamed. I think I am unclear how the streamed solution breaks backwards compatibility - that was definitely not intended.

The aim of my implementation was to decouple historic data formats (such as traditional_pin) in preprocessing and move into a single specification for input, so that internally in Mokapot only one format needs to be handled. In this MR there in a few places we have this internal handling between .tsv and .pin.
I have no hard feelings on either approach, but it this might also conflict with the confidence streaming MR that is upcomming from Elmar. We are aiming for this to be up at the end of the week.

Maybe we can discuss how to best move forward in a call?